ra-tls/libenclave-tls/enclave-tls.c

```diff
@@ -0,0 +1,262 @@
+	wolfTLSv1_2_server_method_func wolfTLSv1_2_server_method;
+	wolfSSL_CTX_free_func wolfSSL_CTX_free;
+	wolfSSL_Cleanup_func wolfSSL_Cleanup;
+	wolfSSL_Debugging_ON_func wolfSSL_Debugging_ON;
+	wolfSSL_Debugging_OFF_func wolfSSL_Debugging_OFF;
+	wolfSSL_Init_func wolfSSL_Init;
+} enclave_tls_function_table;
+
+/* global function table */
+enclave_tls_function_table ftab;
+
+int cert_verify_callback(int preverify, WOLFSSL_X509_STORE_CTX *store)
+{
+	(void) preverify;
+	int ret = 0;
+	if (is_epid_ratls_cert(store->certs->buffer, store->certs->length)) {
+		void *handle = dlopen(LIBENCLAVE_TLS_SGX, RTLD_LAZY);
+		if (NULL == handle) {
+			printf("open error:%s\n", dlerror());
+			return -1;
+		}
+		verify_cert_extensions_func verify_cert_extensions =
+			dlsym(handle, "verify_sgx_cert_extensions");
+		ret = verify_cert_extensions(store->certs->buffer,
+					     store->certs->length);
+		dlclose(handle);
+	}
+
+	fprintf(stderr, "Verifying certificate extensions ... %s\n",
+		ret == 0 ? "Success" : "Failure");
+	return !ret;
+}
+
+void *enclave_tls_init(struct enclave_tls_ctx *ctx)
+{
+	ctab.generate_certificate = &generate_x509;
+
+	if (ctx->quote_type == EPID) {
+		ctx->handle = dlopen(LIBENCLAVE_TLS_SGX, RTLD_LAZY);
+		if (NULL == ctx->handle) {
+			printf("open error:%s\n", dlerror());
+			return -1;
+		}
+#define DLSYM(fn)									\
+		do {									\
+			ftab.fn = dlsym(ctx->handle, #fn "_sgx");				\
+		} while (0)
+
+		DLSYM(wolfSSL_free);
+		DLSYM(wolfSSL_CTX_set_verify);
+		DLSYM(create_key_and_x509);
+		DLSYM(wolfSSL_read);
+		DLSYM(wolfSSL_write);
+		DLSYM(wolfSSL_CTX_new);
+		DLSYM(printf_identity);
+		DLSYM(wolfTLSv1_2_client_method);
+		DLSYM(wolfSSL_X509_get_der);
+		DLSYM(wolfSSL_get_peer_certificate);
+		DLSYM(wolfSSL_connect);
+		DLSYM(wolfSSL_negotiate);
+		DLSYM(wolfSSL_set_fd);
+		DLSYM(wolfSSL_new);
+		DLSYM(wolfTLSv1_2_server_method);
+		DLSYM(wolfSSL_CTX_free);
+		DLSYM(wolfSSL_Cleanup);
+		DLSYM(wolfSSL_Debugging_ON);
+		DLSYM(wolfSSL_Debugging_OFF);
+		DLSYM(wolfSSL_Init);
+#undef DLSYM
+	} else {
+#define DLSYM(fn)                                                                       \
+                do {                                                                    \
+                        ftab.fn = &fn;                                                  \
+                } while (0)
+
+		DLSYM(wolfSSL_free);
+		DLSYM(wolfSSL_CTX_set_verify);
+		DLSYM(create_key_and_x509);
+		DLSYM(wolfSSL_read);
+		DLSYM(wolfSSL_write);
+		DLSYM(wolfSSL_CTX_new);
+		DLSYM(printf_identity);
+		DLSYM(wolfTLSv1_2_client_method);
+		DLSYM(wolfSSL_X509_get_der);
+		DLSYM(wolfSSL_get_peer_certificate);
+		DLSYM(wolfSSL_connect);
+		DLSYM(wolfSSL_negotiate);
+		DLSYM(wolfSSL_set_fd);
+		DLSYM(wolfSSL_new);
+		DLSYM(wolfTLSv1_2_server_method);
+		DLSYM(wolfSSL_CTX_free);
+		DLSYM(wolfSSL_Cleanup);
+		DLSYM(wolfSSL_Debugging_ON);
+		DLSYM(wolfSSL_Debugging_OFF);
+		DLSYM(wolfSSL_Init);
+#undef DLSYM
+	}
+
+	ftab.wolfSSL_Debugging_ON();
```

Comment:
Add a debug parameter in ctx, and then call this if debug is ON.

Reply:
I mean conf structure can add a bool member called debug which is used to enable this function call.